### DISCUSSION TITLE
Retries for jobs with unpredictable duration

### DISCUSSION THREAD

**Discussion by colingavin** [janbjorge/pgqueuer#127] [Ideas]
I'd like to use the job retry mechanism to make my application more robust against workers crashing or being disconnected. However, my jobs have somewhat widely distributed durations, so it's difficult to choose a good `retry_timer` value that doesn't lead to unnecessary retries in some cases.

I'd like to propose a simple way for entrypoints to report that they are still making progress and their job should not be retried. At the API level this would look like:
```
class QueueManager:
  def notify_activity(job_id: models.JobId):
    ...
```
which the entrypoint could call with its job's ID[^1]. All this would do would be to bump the `updated` column for the job to the current time. This would prevent the job from being retried for another `retry_timer` seconds.

A couple enhancements that I could think of on top of this:
- The updates to the `updated` column could be coalesced using a mechanism like `JobBuffer` to reduce query traffic
- The `@QueueManager.entrypoint` decorator could have an optional `heartbeat_interval` argument that, if supplied, would cause the QM to automatically handle calling `notify_activity` while a job was in progress

I'd be happy to prototype this if you agree it'd be worthwhile to have.

[^1]: Note that this is a synchronous method because it needs to be callable from synchronous endpoints. I imagine it would just create a new task within the QM's `TaskManager`. Potentially an async version could also be offered.

**Comment by janbjorge** [comment]
Thank you for the feature suggestion! I agree that adding a notify_activity method to allow entrypoints to report progress would be a valuable enhancement. It would indeed help prevent unnecessary retries for long-running jobs.

Implementing this will require some consideration, especially since synchronous entrypoints run in separate threads, making communication with the main event loop a bit tricky. I'll need to think about the best way to handle this.

I appreciate your offer to prototype the feature. Please feel free to proceed, and we can collaborate on refining it.

EDIT:
After a bit of digging, a detail we need to sort out is the that [updated](https://github.com/janbjorge/pgqueuer/blob/main/pgqueuer/queries.py#L378) column is used to give the _time in queue_ statistics.

**Comment by colingavin** [comment]
Ah those are two good points. I think the thread safety is manageable. I'll have to think a bit about how to handle the "time in queue".

**Reply by janbjorge** [reply]
I agree, i think the best way is to add a new column `heartbeat` that we use insted of the updated column.

**Comment by janbjorge** [comment]
To avoid affecting the time-in-queue statistics, we can introduce a `heartbeat` column (`NOT NULL` with default `NOW()`) in the queue table. The retry logic will then use this `heartbeat` column instead of `updated`.

**Implementation Details:**

- **Add the `heartbeat` column:**

  ```sql
  ALTER TABLE pgqueuer
  ADD COLUMN IF NOT EXISTS heartbeat TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW();
  ```

- **Update the dequeue query to use `heartbeat`:**

  ```sql
  -- In the WHERE clause for retrying picked jobs
  AND ($3::interval IS NOT NULL AND heartbeat < NOW() - $3::interval)
  ```

- **Implement the `notify_activity` method:**

  ```python
  async def notify_activity(self, job_ids: list[models.JobId]) -> None:
      await self.queries.driver.execute(
          f"UPDATE {self.queries.qb.settings.queue_table} "
          f"SET heartbeat = NOW() WHERE id = ANY($1::integer[])",
          job_ids,
      )
  ```

Entrypoints can call `notify_activity([job.id])` whenever they need to report activity.

**Comment by colingavin** [comment]
That definitely seems like the best way to go. I think in the dequeue query we would also need to update the heartbeat:
```sql
SET status = 'picked', updated = NOW(), heartbeat = NOW()
```
It may have already been some time since the job was originally inserted, so the retry timeout should be reset.

For the thread-safety issue, I believe that since you're using `anyio.to_thread.run_sync` to run synchronous entrypoints, we can simply have the synchronous version of the `notify_activity` method use `anyio.from_thread.run_sync` to call back into the event loop.

**Reply by janbjorge** [reply]
> For the thread-safety issue, I believe that since you're using anyio.to_thread.run_sync to run synchronous entrypoints, we can simply have the synchronous version of the notify_activity method use anyio.from_thread.run_sync to call back into the event loop.

I think if we use the buffer in the main loop to reduce db calls, then we are fine wrt. potential threading issues. I want to avoid lib. code doing io-calls from any other places then the main thread.

**Comment by janbjorge** [comment]
Working on: https://github.com/janbjorge/pgqueuer/pull/128 hopefully i think it will be useful to track the healthcheckins.

**Comment by janbjorge** [comment]
@colingavin Could you take a look a / try this PR? https://github.com/janbjorge/pgqueuer/pull/129
